core/src/main/java/com/sequenceiq/cloudbreak/service/template/ClusterTemplateService.java

```diff
@@ -337,21 +337,25 @@ private String getTemplateString(ClusterTemplate clusterTemplate) {
         }
     }
 
-    private void updateDefaultClusterTemplates(Workspace workspace) {
+    private void updateDefaultClusterTemplatesIfNeed(Workspace workspace) {
         Set<ClusterTemplate> clusterTemplates = clusterTemplateRepository.findAllByNotDeletedInWorkspace(workspace.getId());
         if (clusterTemplateLoaderService.isDefaultClusterTemplateUpdateNecessaryForUser(clusterTemplates)) {
-            LOGGER.debug("Modifying clusterDefinitions based on the defaults for the '{} ({})' workspace.", workspace.getName(), workspace.getId());
-            Collection<ClusterTemplate> outdatedTemplates = clusterTemplateLoaderService.collectOutdatedTemplatesInDb(clusterTemplates);
-            LOGGER.debug("Outdated clusterDefinitions collected: '{}'.", outdatedTemplates.size());
-            delete(new HashSet<>(outdatedTemplates));
-            LOGGER.debug("Outdated clusterDefinitions deleted: '{}'.", outdatedTemplates.size());
-            clusterTemplates = clusterTemplateRepository.findAllByNotDeletedInWorkspace(workspace.getId());
-            LOGGER.debug("None deleted clusterDefinitions collected: '{}'.", clusterTemplates.size());
-            clusterTemplateLoaderService.loadClusterTemplatesForWorkspace(clusterTemplates, workspace, this::createAll);
-            LOGGER.debug("ClusterDefinition modifications finished based on the defaults for '{}' workspace.", workspace.getId());
+            updateDefaultClusterTemplatesWithSync(workspace, clusterTemplates);
         }
     }
 
+    private synchronized void updateDefaultClusterTemplatesWithSync(Workspace workspace, Set<ClusterTemplate> clusterTemplates) {
```

Comment:
How does it solve the problem when more than 1 CB instance is running? (eg 3 pods in prod) I think we can solve the problem with transactionservice.required().

Reply:
I'm using the sync block and added a try-catch to handle the `OptimisticLockingFailureException`. I'm not preferred the transaction service, because it ignores all of the exceptions and I'm afraid we will hide some error what we don't want